src/linked_hash_map.rs

```diff
@@ -0,0 +1,509 @@
+        unsafe{ (*self.k).eq(&*other.k) }
+    }
+}
+
+impl<K: Eq> Eq for KeyRef<K> {}
+
+impl<K, V> LinkedHashMapEntry<K, V> {
+    fn new(k: K, v: V) -> LinkedHashMapEntry<K, V> {
+        LinkedHashMapEntry {
+            key: k,
+            value: v,
+            next: ptr::null_mut(),
+            prev: ptr::null_mut(),
+        }
+    }
+}
+
+impl<K: Hash + Eq, V> LinkedHashMap<K, V> {
+    /// Create a linked hash map.
+    pub fn new() -> LinkedHashMap<K, V> {
+        let map = LinkedHashMap {
+            map: HashMap::new(),
+            head: unsafe{ mem::transmute(box mem::uninitialized::<LinkedHashMapEntry<K, V>>()) },
+        };
+        unsafe {
+            (*map.head).next = map.head;
+            (*map.head).prev = map.head;
+        }
+        return map;
+    }
+
+    /// Inserts a key-value pair into the map. If the key already existed, the old value is
+    /// returned.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map = LinkedHashMap::new();
+    ///
+    /// map.insert(1i, "a");
+    /// map.insert(2, "b");
+    /// assert_eq!(map.get(&1), Some(&"a"));
+    /// assert_eq!(map.get(&2), Some(&"b"));
+    /// ```
+    pub fn insert(&mut self, k: K, v: V) -> Option<V> {
+        let (node_ptr, node_opt, old_val) = match self.map.get_mut(&KeyRef{k: &k}) {
+            Some(node) => {
+                let old_val = mem::replace(&mut node.value, v);
+                let node_ptr: *mut LinkedHashMapEntry<K, V> = &mut **node;
+                (node_ptr, None, Some(old_val))
+            }
+            None => {
+                let mut node = box LinkedHashMapEntry::new(k, v);
+                let node_ptr: *mut LinkedHashMapEntry<K, V> = &mut *node;
+                (node_ptr, Some(node), None)
+            }
+        };
+        match node_opt {
+            None => {
+                // Existing node, just update LRU position
+                self.detach(node_ptr);
+                self.attach(node_ptr);
+            }
+            Some(node) => {
+                let keyref = unsafe { &(*node_ptr).key };
+                self.map.insert(KeyRef{k: keyref}, node);
+                self.attach(node_ptr);
+            }
+        }
+        old_val
+    }
+
+    /// Return a value corresponding to the key in the map.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map = LinkedHashMap::new();
+    ///
+    /// map.insert(1i, "a");
+    /// map.insert(2, "b");
+    /// map.insert(2, "c");
+    /// map.insert(3, "d");
+    ///
+    /// assert_eq!(map.get(&1), Some(&"a"));
+    /// assert_eq!(map.get(&2), Some(&"c"));
+    /// ```
+    pub fn get(&self, k: &K) -> Option<&V> {
+        self.map.get(&KeyRef{k: k}).map(|e| &e.value)
+    }
+
+    /// Return a value corresponding to the key in the map.
+    ///
+    /// If value is found, it is pushed to the end of the list.
```

Comment:
I'm not sure `end` is a meaningful metaphor here? Are we providing a logical stack or queue? We should be using that terminology, I think.